### PR TITLE
[#12808][feat] AutoDeploy: Custom attn mask support for attention backends

### PR DESCRIPTION
**Pull request by bmarimuthu-nv**
## Description

This branch adds custom attention mask injection infrastructure for the AutoDeploy pipeline, specifically targeting Gemma4's VLM attention patterns. The
   work spans:                                                                                                                                            
  1. A new transform (inject_custom_attention_mask) with a provider registry pattern                                                                      
  2. A new get_dynamic_inputs() hook in the AttentionDescriptor ABC to forward runtime tensors through the cached attention op                            
  3. A default extra-arg factory system on SequenceInfo for warmup/CUDA-graph passes                                                                      
  4. A new Triton kernel (_paged_context_masked_kernel) for masked prefill                                                                                
  5. Mask support plumbed into torch and triton_paged backends                                                                                            
  6. Refactoring of layout extraction to use extract_op_args() consistently 

This is infra only. A follow up Gemma4 multimodal PR PR will implement the end 2 end for the model.

# Custom Attention Mask Flow

## Key Design Points

- Mask is computed **outside** the graph (no mask-building ops in the exported graph). In-graph mask computation hits symbolic shape and device errors after graph recompilation during KV cache transforms
- The graph only receives a `custom_attn_mask` placeholder (defaults to `None`)
- Provider registry maps `(model_type, backend)` → mask wiring function
- `get_dynamic_inputs()` on `AttentionDescriptor` ensures the mask tensor flows through the cached attention op (between caches and constants)
- Triton kernel requires **broadcast mask** `[B, 1, S_q, S_k]`; torch backend supports `[B, N, S_q, S_k]` via slicing
- custom_attn_mask should not be expected to carry sliding-window logic. The backend must apply sliding window on top of whatever user mask is provided.

## Block Diagram

```
┌─────────────────────────────────────────────────────────────────────────┐
│                        TRANSFORM TIME (graph rewrite)                   │
│                                                                         │
│  InferenceOptimizer                                                     │
│    │                                                                    │
│    ▼                                                                    │
│  InjectCustomAttentionMask transform                                    │
│    │                                                                    │
│    ├─ infer_model_type(factory) ──► "gemma4"                           │
│    │                                                                    │
│    ├─ AttentionMaskProviderRegistry.get("gemma4", backend)             │
│    │     │                                                              │
│    │     ▼                                                              │
│    │   _gemma4_*_mask_provider(ctx, attn_node)                         │
│    │     │                                                              │
│    │     ▼                                                              │
│    │   ctx.add_or_retrieve_input("custom_attn_mask", val=None)         │
│    │     │                  ▲                                           │
│    │     │                  │ (cached — built once, shared across       │
│    │     │                  │  all attn nodes in the graph)             │
│    │     ▼                                                              │
│    │   placeholder node: custom_attn_mask                              │
│    │                                                                    │
│    ├─ For each torch_attention node:                                    │
│    │     set attn_mask arg ──► placeholder node                        │
│    │                                                                    │
│    ▼                                                                    │
│  _InsertCachedOperator (kvcache.py)                                    │
│    │                                                                    │
│    ├─ get_dynamic_inputs(attn_node) ──► [attn_mask node]               │
│    │                                                                    │
│    ▼                                                                    │
│  cached_attention_op(q, k, v, *meta, *caches, *dynamic, *constants)    │
│                                            ▲                            │
│                                            │                            │
│                              custom_attn_mask inserted here             │
└─────────────────────────────────────────────────────────────────────────┘

┌─────────────────────────────────────────────────────────────────────────┐
│                        RUNTIME (forward pass)                           │
│                                                                         │
│  VLM Wrapper (outside the graph)                                        │
│    │                                                                    │
│    ├─ Prefill with vision tokens?                                       │
│    │     YES ──► compute mask [B, 1, S_q, S_k] from token_type_ids     │
│    │     NO  ──► pass None (decode / text-only)                        │
│    │                                                                    │
│    ▼                                                                    │
│  graph(input_ids, position_ids, custom_attn_mask=mask_or_none)         │
│    │                                                                    │
│    ▼                                                                    │
│  ┌───────────────────────────────────────────────────────┐              │
│  │  Backend dispatch (per attention layer)                │              │
│  │                                                       │              │
│  │  if custom_attn_mask is None:                         │              │
│  │    ├─ torch_backend ──► standard causal kernel         │              │
│  │    └─ triton_paged  ──► triton_paged_context()        │              │
│  │                                                       │              │
│  │  if custom_attn_mask is not None:                     │              │
│  │    ├─ torch_backend ──► _torch_context_mha()          │              │
│  │    │    mask[idx, :, :S_q, :S_kv] ──► masked_fill_   │              │
│  │    │                                                  │              │
│  │    └─ triton_paged  ──► _paged_context_masked_kernel  │              │
│  │         mask[B, 0, q_pos, kv_pos] ──► per-page load  │              │
│  │         (broadcast over heads, head dim must be 1)    │              │
│  └───────────────────────────────────────────────────────┘              │
│                                                                         │
│  Warmup / CUDA-graph capture passes:                                    │
│    SequenceInfo.nest_sequences()                                        │
│      │                                                                  │
│      ├─ _default_extra_arg_factories["custom_attn_mask"](info)         │
│      │    ──► returns None (no mask during warmup)                      │
│      │                                                                  │
│      └─ per-request extra_args override if present                      │
└─────────────────────────────────────────────────────────────────────────┘
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Summary by CodeRabbit

# Release Notes

* **New Features**
  * Added custom attention mask support across multiple attention backends (torch, triton_paged, torch_attention). Introduced an attention mask provider registry enabling model-specific mask handling. Enhanced graph transformation pipeline with mask injection capabilities. Extended attention descriptors to support dynamic inputs and shared KV caching.

* **Tests**
  * Added comprehensive test coverage for custom attention mask functionality and provider registry integration.

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/custom_ops/attention/triton_paged_attention.py [hunk 1]
```diff
@@ -209,7 +209,7 @@ def _get_num_splits(max_seq_len: int, batch_size: int, n_kv_heads: int, page_siz
         triton.Config({}, num_warps=8, num_stages=2),
         triton.Config({}, num_warps=8, num_stages=3),
     ],
-    key=["HEAD_DIM", "PAGE_SIZE", "HEAD_RATIO_PADDED"],
+    key=["HEAD_DIM", "PAGE_SIZE", "HEAD_RATIO_PADDED", "SLIDING_WINDOW"],
```
Comment (bmarimuthu-nv) [inline tensorrt_llm/_torch/auto_deploy/custom_ops/attention/triton_paged_attention.py]: Sliding window changes pulled from https://github.com/NVIDIA/TensorRT-LLM/pull/12710/changes/141e0d467f376d05df6d4d438d6ca9d9ef399312#diff-e5e901759fe5eb0bfbd8a6fbbdbe5ed498a49b93bff19830e9c95c10745d6b14 and updated to support user custom_attn_mask + sliding_window on top

**Review by lucaslie — COMMENTED** [review]
## Design Review: Custom Attention Mask Infrastructure

Thanks for the work here — several components are valuable and should be kept regardless of the overall architecture. Specifically:

- **Sliding window in triton_paged** (both context and decode kernels) is a solid improvement
- **`extract_op_args` refactoring** replacing ad-hoc positional arg extraction is clean
- **`_paged_context_masked_kernel`** Triton kernel for masked prefill will be needed regardless of design
- **`_torch_context_mha_readonly`** for read-only context attention is useful

However, the overall design has several structural issues that will make it fragile at scale. Below is a detailed analysis and an alternative proposal.

---

### Critical Issues

#### 1. Type Variance for Graph Inputs

The PR passes `custom_attn_mask` as a graph input that alternates between `None` (decode/warmup/text-only) and a tensor (VLM prefill). This is problematic because:

- `torch.export` requires tensor-typed graph inputs; `None` is not a valid dynamic shape source.
- CUDA graph capture requires fixed tensor addresses and shapes. The monolithic decode path uses `CapturedGraph` which hashes static inputs; switching between `None` and a tensor changes the hash, causing fallback to eager execution.

#### 2. `_default_extra_arg_factories` Is Over-Engineered

The new factory system on `SequenceInfo` is unnecessary because:

- The existing `capture_forward_kwargs` mechanism in `export_to_gm.py` already captures whatever kwargs the submodule receives during export, including additional inputs.
- Monolithic decode warmup uses `set_generate_only_batch()` which doesn't pass multimodal extras.
- Piecewise prefill warmup in `_setup_piecewise_mixed_batch` also uses synthetic sequences without extras.

#### 3. Provider Registry Creates Tight Coupling

The `AttentionMaskProviderRegistry` keyed by `(model_type, backend)` means:

- Every new model needing masks must register providers for **every** backend.
- Model-specific mask semantics are coupled to backend code.
- Adding a new backend requires updating all model providers.

#### 4. `mutates_args` Regression

`torch_cached_attention_with_cache` changed from `mutates_args=("k_cache", "v_cache")` to `mutates_args=()`. The k/v caches **are** mutated in-place during context attention (`_torch_context_mha` writes to them). Removing this annotation means the graph compiler may not track these mutations correctly.

#### 5. Mask Computation Outside Graph Is Fragile

Computing the mask in the VLM wrapper (outside the graph) means:
- The wrapper must know backend-specific mask formats (e.g., triton requires `[B, 1, S_q, S_k]` uint8).
- The mask must be constructed before the graph call, requiring access to runtime metadata.
- The graph cannot be tested in isolation for mask correctness.

---

### Alternative Design Proposal

#### Principle: Two Modes for `attn_mask`, Analogous to Attention's Own Cached/Uncached Swap Pattern

The canonical IR op `torch_attention` already has an `attn_mask` parameter. Instead of building a separate injection transform, make better use of this parameter with two modes. The existing `is_causal` boolean and `attn_mask: Optional[torch.Tensor]` parameters are preserved as-is.

#### Mode A: String Mask Types (Design Now, Implement Later)

In addition to `Optional[torch.Tensor]`, allow `attn_mask` to accept well-known string literals (e.g. `"causal"`, `"bidirectional"`). Each attention backend maps recognized strings to its optimized implementation or raises an error if unsupported. The existing `is_causal` boolean is kept; when `attn_mask` is a string, it takes precedence.

This mode is not required for the immediate VLM iteration — it should be part of the design but can be implemented as a follow-up.

#### Mode B: Mask Custom Op (for VLM/Gemma4)

For complex, model-specific masks that cannot be described by a string:

**Key layout constraint**: During export, the model runs in `[B, S]` layout (e.g., `set_example_sequence` creates `[2, 4]` inputs). Only after the kvcache transform does the graph switch to ragged `[1, total_tokens]` layout. The mask custom op's reference implementation must therefore operate in `[B, S]` layout.

1. **Define a mask custom op in model code** (not backend code):

```python
@torch.library.custom_op("auto_deploy::gemma4_build_vlm_mask", mutates_args=())
def gemma4_build_vlm_mask(
    is_vision_token: torch.Tensor,  # [B, S] bool - same layout as input_ids at export time
    seq_len: int,
) -> torch.Tensor:
    # Reference: build VLM attention mask from vision token indicator.
    # Operates in [B, S] layout (pre-kvcache-transform).
    # Returns mask in [B, 1, S, S] compatible with torch_attention attn_mask.
    ...
```

2. **During export**: The model's forward calls this custom op to build the mask from `is_vision_token`, passes the result to `torch_attention(attn_mask=...)`. The existing `capture_forward_kwargs` mechanism naturally captures `is_vision_token` as a graph input with dynamic shapes. No new infrastructure needed.

3. **During kvcache transform**: Just like `torch_attention` is swapped for `cached_attention`, the mask custom op is identified as a single node in the graph and swapped for a backend-specific version. The backend version:
   - Receives the original args (e.g., `is_vision_token` — now in ragged layout)
   - Receives metadata args (e.g., `batch_info_host`, sequence boundaries)
   - Produces a mask in the format the backend's cached attention kernel expects

4. **Backend-specific mask functions registered from model code**:

```python
# In model code (e.g., _torch/auto_deploy/models/gemma4.py), not backend code
MaskOpRegistry.register(
    source_op="auto_deploy::gemma4_build_vlm_mask",
    backend="triton_paged",
)(gemma4_triton_paged_mask_fn)
```

The registration is model-driven: each model declares which backends it supports masks for. A backend with no registered mask function for a given mask op raises a clear error. This avoids the `(model_type, backend)` cartesian product problem.

#### VLM Prefill/Decode Handling

Following the existing prefill/decode specialization pattern:

- **Prefill with images**: Backend mask function reads `is_vision_token` (ragged layout), uses metadata (`batch_info_host`, `cu_seqlen`) to build the VLM mask.
- **Prefill without images**: `is_vision_token` is all-False; backend mask function produces standard causal mask or returns a sentinel for the fast path.
- **Decode**: Backend mask function checks `batch_info_host`, sees decode-only, returns a no-op. This is CUDA-graph compatible because:
  - Decode runs monolithic CUDA graphs.
  - The mask function still executes (same graph structure) but produces a dummy output.
  - The attention kernel's decode path does not consume the mask output.
  - `is_vision_token` is padded to the CUDA graph bucket size (all-False), so tensor shapes are stable.

The mask computation itself is **not** expected to be CUDA-graph compatible (it may involve data-dependent control flow). This is fine because mask computation only matters for prefill, and prefill is **not** CUDA-graphed. For decode CUDA graphs, the mask function runs but is a no-op.

#### Key Advantages

1. **No type variance**: `is_vision_token` is always a tensor. No `None`/tensor switching for graph inputs.
2. **No `_default_extra_arg_factories`**: Uses existing `capture_forward_kwargs` and `extra_args` mechanisms.
3. **Separation of concerns**: Model code defines mask semantics and registers backend implementations. Backend attention code only needs to accept the mask output.
4. **CUDA graph compatible by design**: Decode ignores mask output; mask function is a no-op.
5. **Testable**: Mask computation is in the graph — graph-level tests verify correctness end-to-end.
6. **Preserves existing API**: `is_causal` kept as-is; `attn_mask` extended naturally.

---

### Recommendation

I'd suggest splitting the valuable parts (sliding window support, `extract_op_args` cleanup, the triton masked kernel) into separate PRs, and redesigning the mask injection infrastructure along the lines above.

**Comment by bmarimuthu-nv** [comment]
Covered in https://github.com/NVIDIA/TensorRT-LLM/pull/12861
